doc/whats_new/v0.20.rst

```diff
@@ -51,6 +51,13 @@ Changelog
   avoid pickling errors caused by the serialization of their methods.
   :issue:`12171` by :user:`Thomas Moreau <tomMoral>`.
 
+:mod:`sklearn.metrics`
+........................
+
+- |Fix| Fixed a bug in :func:`metrics.silhouette_score` where diagonal entries
+  of a precomputed distance matrix affected the final score.
```

Comment:
It would have affected non-precomputed too, wouldn't it, if there were precision issues in calculating those diagonals? We should handle that case too